File: done/70 - climbStairs.py
#https://leetcode.com/problems/climbing-stairs/

# Approach 1: Bottom up
class Solution:
    def climbStairs(self, n: int) -> int:
        if n <= 3:
            return n
        n1, n2 = 2, 3

        for i in range(4, n + 1):
            temp = n1 + n2
            n1 = n2
            n2 = temp
        return n2
    
# Approach 2: Memoization (top down)
class Solution:
    def climbStairs(self, n: int) -> int:
        cache = {}
        def memo(i):
            if i > n:
                return 0
            if i in cache:
                return cache[i]
            if i == n: 
                return 1

            cache[i] = memo(i + 1) + memo(i + 2)
            return cache[i]

        return memo(0)

File: DSA/done/70 - climbStairs.cpp
// https://leetcode.com/problems/climbing-stairs/

#include <vector>

// Memo
class Solution {
private:
    int solve(int currentStep, int targetStep, std::vector<int>& memo) {
        // Base case: If we reach exactly the top, we found 1 valid way
        if (currentStep == targetStep) {
            return 1;
        }
        // Base case: If we overstep the top, this path is invalid
        if (currentStep > targetStep) {
            return 0;
        }
        
        // If we have already calculated the ways for this step, return the cached result
        if (memo[currentStep] != -1) {
            return memo[currentStep];
        }
        
        // Recursively find ways by taking 1 step + taking 2 steps
        int takeOneStep = solve(currentStep + 1, targetStep, memo);
        int takeTwoSteps = solve(currentStep + 2, targetStep, memo);
        
        // Cache and return the result for the current step
        return memo[currentStep] = takeOneStep + takeTwoSteps;
    }

public:
    int climbStairs(int n) {
        // Initialize memo vector with -1 to indicate uncalculated steps
        // Size is n + 1 to comfortably index from 0 up to n
        std::vector<int> memo(n + 1, -1);
        
        // Start from step 0
        return solve(0, n, memo);
    }
};

// Tabular
class Solution {
public:
    int climbStairs(int n) {
        int prev{1}, prev2{0};
        for (int i{n - 1}; i >= 0; i--) {
            int cur{prev + prev2};
            prev2 = prev;
            prev = cur;
        }

        return prev;
    }
};